<a href="https://colab.research.google.com/github/arshdeepbangar/AAI2025/blob/main/Arsh_Bangar_Exercise_2_ReACT_Code_Generation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2 - Code Generation with ReACT-Style Prompting

**Goal:** Use an iterative cycle: plan -> generate code -> run -> observe -> fix.

**Tools used:** ChatGPT for code-generation prompts and Google Colab/Python for execution.

**Coding task:** Create a Python function named `analyze_orders` that receives a list of order amounts and returns count, total, average, and highest valid amount.

## Attempt 1 - Vague prompt

**Prompt:** Write Python code that analyzes a list of order amounts and gives me the count, total, average, and highest amount.

This first prompt is intentionally simple so it can be tested and improved.

In [1]:
def analyze_orders_v1(order_amounts):
    amounts = [float(x) for x in order_amounts if x is not None and float(x) > 0]
    return {
        "count": len(amounts),
        "total": round(sum(amounts), 2),
        "average": round(sum(amounts) / len(amounts), 2),
        "highest": round(max(amounts), 2),
    }

test_orders = [120.50, 75, "bad", -10, 250.25, None, 89.99]
try:
    print(analyze_orders_v1(test_orders))
except Exception as e:
    print(f"Observed error: {type(e).__name__}: {e}")

Observed error: ValueError: could not convert string to float: 'bad'


## Observation from the first run

The first version fails on the string `"bad"`. The prompt did not explain how invalid values should be handled, so the generated code was not robust enough.

## Refined ReACT-style prompt

**Stage 1 - Plan:** Give a brief 3-bullet plan describing the algorithm and important edge cases. Keep the plan concise; do not provide hidden chain-of-thought.

**Stage 2 - Act:** Generate Python code for `analyze_orders(order_amounts)`.

Requirements:
- Input may contain integers, floats, numeric strings, invalid strings, `None`, zero, or negative values.
- Keep only values that can be converted to a number and are greater than 0.
- Return a dictionary with keys `count`, `total`, `average`, and `highest`.
- Round `total`, `average`, and `highest` to 2 decimal places.
- If there are no valid positive amounts, raise `ValueError("No valid order amounts")`.
- Use only the Python standard library.
- Keep the code easy to read.

**Stage 3 - Run:** Test with `[120.50, 75, "bad", -10, 250.25, None, 89.99]`.

**Stage 4 - Observe:** Review the exact output or error.

**Stage 5 - Fix:** If the result is wrong or an error occurs, revise the code and briefly state what changed.

### AI plan produced from the refined prompt

- Loop through each item and safely try to convert it to a number.
- Keep only positive numeric values, then calculate the requested statistics.
- Raise a clear `ValueError` when no valid positive amounts remain.

In [ ]:
def analyze_orders(order_amounts):
    valid_amounts = []
    for value in order_amounts:
        try:
            amount = float(value)
        except (TypeError, ValueError):
            continue
        if amount > 0:
            valid_amounts.append(amount)

    if not valid_amounts:
        raise ValueError("No valid order amounts")

    total = sum(valid_amounts)
    return {
        "count": len(valid_amounts),
        "total": round(total, 2),
        "average": round(total / len(valid_amounts), 2),
        "highest": round(max(valid_amounts), 2),
    }

orders = [120.50, 75, "bad", -10, 250.25, None, 89.99]
print("Successful result:", analyze_orders(orders))

try:
    analyze_orders([None, "bad", -2, 0])
except ValueError as e:
    print("Expected edge-case message:", e)

Successful result: {'count': 4, 'total': 535.74, 'average': 133.94, 'highest': 250.25}
Expected edge-case message: No valid order amounts


## Observation after the fix

The revised code runs successfully. Invalid entries are skipped, negative and zero values are ignored, valid values are calculated correctly, and the all-invalid test raises the required error message.

## Fix/iteration note

The refined prompt added exact input rules, output keys, rounding, allowed libraries, and error handling. This completes the ReACT-style cycle: **plan -> act -> run -> observe -> fix**.

## Final successful output

For the main test, the function returns:

`{'count': 4, 'total': 535.74, 'average': 133.94, 'highest': 250.25}`

The edge-case test prints: `No valid order amounts`.